controllers/helmrelease_controller_release.go

```diff
@@ -0,0 +1,426 @@
+		r.patchStatus(ctx, hr)
+	// The state has not changed and the release is not in a failed state.
+	case rls != nil && remediation.GetFailureCount(*hr) == 0:
+		v2.HelmReleaseReady(hr, rls.Info.Description)
+		return ctrl.Result{RequeueAfter: hr.Spec.Interval.Duration}, nil
+	// We have exhausted our retries.
+	case remediation.RetriesExhausted(*hr):
+		v2.HelmReleaseNotReady(hr, meta.ReconciliationFailedReason, "exhausted release retries")
+		return ctrl.Result{RequeueAfter: hr.Spec.Interval.Duration}, nil
+	// Our previous remediation attempt failed, skip release to retry.
+	case hr.Status.LastSuccessfulReleaseRevision > 0 && hr.Status.LastReleaseRevision != hr.Status.LastSuccessfulReleaseRevision:
+		return ctrl.Result{RequeueAfter: hr.Spec.Interval.Duration}, nil
+	}
+
+	// Attempt to download and load the chart. As we already checked
+	// if an artifact is advertised in the chart object, any error
+	// we encounter is expected to be transient.
+	// This must always be done right before making the actual release,
+	// as depending on the chart this can be a (memory) expensive
+	// operation.
+	loadedChart, err := r.loadHelmChart(hc)
+	if err != nil {
+		v2.HelmReleaseNotReady(hr, v2.ArtifactFailedReason, err.Error())
+		r.event(hr, hr.Status.LastAttemptedRevision, events.EventSeverityError, err.Error())
+		return ctrl.Result{}, err
+	}
+
+	// Register our new release attempt and make the release.
+	v2.HelmReleaseAttempt(hr, hc.GetArtifact().Revision, valuesChecksum)
+	err = makeRelease(*hr, loadedChart, values)
+
+	// Always record the revision when a new release has been made.
+	hr.Status.LastReleaseRevision = util.ReleaseRevision(run.GetLastPersistedRelease())
+
+	if err != nil {
+		// Record the release failure and increment the failure count.
+		meta.SetResourceCondition(hr, v2.ReleasedCondition, metav1.ConditionFalse, failureReason, err.Error())
+		v2.HelmReleaseNotReady(hr, failureReason, err.Error())
+		remediation.IncrementFailureCount(hr)
+		r.event(hr, hr.Status.LastAttemptedRevision, events.EventSeverityError, err.Error())
+
+		// We should still requeue on the configured interval so that
+		// both the Helm storage and ValuesFrom references are observed
+		// on an interval.
+		return ctrl.Result{RequeueAfter: hr.Spec.Interval.Duration}, nil
+	}
+
+	meta.SetResourceCondition(hr, v2.ReleasedCondition, metav1.ConditionTrue, successReason,
+		run.GetLastPersistedRelease().Info.Description)
+	r.event(hr, hr.Status.LastAttemptedRevision, events.EventSeverityInfo, run.GetLastPersistedRelease().Info.Description)
+
+	return ctrl.Result{RequeueAfter: hr.Spec.Interval.Duration}, nil
+}
+
+func (r *HelmReleaseReconciler) reconcileTest(ctx context.Context, log logr.Logger, run *runner.Runner, hr *v2.HelmRelease) (ctrl.Result, error) {
+	// If this release was already marked as successful,
+	// we have nothing to do.
+	if hr.Status.LastReleaseRevision == hr.Status.LastSuccessfulReleaseRevision {
+		return ctrl.Result{}, nil
```

Comment:
Shouldn't we always be setting `requeueAfter: hr.Spec.Interval.Duration` here and elsewhere when there is nothing to do, to ensure we don't end up using whatever the default value is? Or alternatively should we override to that duration after running each step if there were no errors and the requeueAfter is zero?

Reply:
The latter has my preference I think.